design-proposals/klip-1-improve-udf-interfaces.md

```diff
@@ -0,0 +1,329 @@
+
+> ```
+> Name        : FOO
+> Author      : Confluent
+> Type        : scalar
+> Jar         : internal
+> Variations  :
+>
+>       Variation   : FOO(ARRAY<ANY<T>>, ANY<U>)
+>       Returns     : ANY<T>
+> ```
+
+### Complex Aggregation
+
+* The `UDAF` section in `udf.rst` will update the example from `Sum` to `Average` and include
+non-trivial usage of the `reduce` API:
+
+> The class below creates a UDAF named ``my_average``...
+>
+> .. code:: java
+> 
+>```java
+>@UdafDescription(name="my_average", description="averages")
+>public class AverageUdaf {
+>  
+>  private static final Schema SCHEMA = SchemaBuilder.struct()
+>                                         .field("sum", Schema.INT64_SCHEMA)
+>                                         .field("count", Schema.INT32_SCHEMA)
+>                                         .build();
+>
+>  @UdafFactory(description = "averages longs")
+>  public static TableUdaf<Long, Struct, Double> createAverageLong() {
+>    return new TableUdaf<Long, Struct, Double>() {
+>      @Override
+>      public Struct initialize() { 
+>        return new Struct(SCHEMA).put("sum", 0L).put("count", 0); 
+>      }
+>      
+>      @Override
+>      public Struct aggregate(final Long value, final Struct aggregate) {
+>        return new Struct(SCHEMA)
+>                  .put("sum", aggregate.get("sum") + value)
+>                  .put("count", aggregate.get("count") + 1);
+>      }
+>      
+>      @Override
+>      public Struct undo(final Long valueToUndo, final Struct aggregate) {
+>        return new Struct(SCHEMA)
+>                  .put("sum", aggregate.get("sum") - valueToUndo)
+>                  .put("count", aggregate.get("count") - 1);
+>      }
+>      
+>      @Override
+>      public Struct merge(final Struct aggOne, final Struct aggTwo) {
+>        return new Struct(SCHEMA)
+>                  .put("sum", aggOne.get("sum") + aggTwo.get("sum"))
+>                  .put("count", aggOne.get("count") + aggTwo.get("count"));
+>      }
+>      
+>      @Override
+>      public Double reduce(final Struct aggregate) {
+>        return ((Double) aggregate.getInt64("sum")) / aggregate.getInt32("count");
+>      }
+>    };
+>  }
+>}
+>```
+
+* For exportable UDAFs, we will also need to generate updated documentation for `DESCRIBE FUNCTION`
+calls to make sure that the output value is not the intermediate value, but rather the final 
+exported value. For example:
+
+> ```
+> ksql> DESCRIBE FUNCTION AVERAGE;
+> 
+> Name        : AVERAGE
+> Author      : Confluent
+> Overview    : Returns the average value for a window.
+> Type        : aggregate
+> Jar         : internal
+> Variations  :
+> 
+>	Variation   : AVERAGE(BIGINT)
+>	Returns     : DOUBLE
+> // Note that the signature for this is actually Udaf<Long, Struct, DOUBLE> as seen above
+>```
+
+* Syntax reference needs to be updated to reflect any new UDF/UDAFs that are implemented as part of
+this KLIP.
```

Comment:
Perhaps worth adding another note about the compatibility-breaking change around the new UDAF structure introduced by complex aggregations. (It's already mentioned above, but I think it'd be good to call it out again here.) Either a note here in the docs-update section about updating the upgrade notes, or a mention in the "Compatibility implications" section below.